File: java/07.07.19/nokia.java
import java.util.*;
public class nokia
{
    static String[] codes = {".", "abc", "def", "ghi", "jklm", "no", "pqrs", "tu", "vwxy", "z"};
    public static void step(String n, String ans)
    {
        if(n.length() == 0)
        {
            System.out.println(ans);
            return;
        }

        char firstChar = n.charAt(0);
        String restOfN = n.substring(1, n.length());
        String code = codes[firstChar - '0'];
        for(int i = 0; i < code.length(); i++)
        {
            step(restOfN, ans + code.charAt(i));
        }
        return;
    }

    public static void main(String[] argc)
    {
        Scanner scn = new Scanner(System.in);
        String n = scn.nextLine();
        step(n, "");
    }
}

File: cpp/05.07.19/nokia.cpp
#include <iostream>
#include <String>
#include <Vector>

using namespace std;

string codes[10] = {".", "abc", "def", "ghi", "jklm", "no", "pqrs", "tu", "vwxy", "z"};

vector<string> pattern(string s)
{
    if(s.length() == 0)
    {
        vector<string> result;
        result.push_back("");
        return result;
    }
    
    char firstKey = s[0];
    string nextInput = s.substr(1, s.length() - 1);
    vector<string> smallerPattern = pattern(nextInput);
    vector<string> result;
    int code = firstKey - '0';
    for(int i = 0; i < codes[code].length(); i++)
    {
        char currentNumberAlphabet = codes[code][i];
        for(int j = 0; j < smallerPattern.size(); j++)
        {
            string elementOfSmallerPattern = smallerPattern[j];
            result.push_back(currentNumberAlphabet + elementOfSmallerPattern);
        }
    }
    return result;
}

int main(int argv, char** argc)
{
    string keys;
    cin >> keys;
    vector<string> ans = pattern(keys);
    for(int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << endl;
    }
}